### ISSUE TITLE
Naming conventions for the project

### ISSUE THREAD

**Issue by yipengsun** [umd-lhcb/lhcb-ntuples-gen#39]
In this issue, we want to define the following naming conventions of this project:

1. Particle names in the branches of ntuples
2. Names of the generated ntuples (incl. folders)
3. Names of the raw `.dst` files (incl. folders)
4. Name of the `reco_*.py` files (incl. folders)

We will document this in our wiki. Also, we'd like to automate this as much as possible (checker, etc.) so that it's easier to follow once one get used to this project.

**Comment by manuelfs** [comment]
Yipeng made [proposals](https://umd-lhcb.github.io/lhcb-ntuples-gen/ntupling/nomenclature/) for naming files and folders

### File names
Proposal is `<date>-<particle_name>-<reconstruction_mode>-<year>-<polarity>-<additional_flags>.root`. What is <particle_name>? The examples given are `Dst` and `D0`, so is this the reconstructed sample? So other samples would be WSmu, WSpi and so on?

My comments:
- Do we need the date in the file names? I wonder if it would be annoying to tab complete, but it may be useful
- Is it more useful to start with the date or with the sample?
- For me it is much easier to read the date if we write it as YY_MM_DD
- I think it is better to simply copy the DaVinci path for the file names. The names would be longer but this would be automatic naming and it would be more directly useful to know which sample it was. 
  - The `/` would be substituted with `_`, so the fields would be separated with `--`

An example name with my suggestions may look as `Dst--20_05_08--MC_2011_11874091_Beam3500GeV-2011-MagDown-Nu2-Pythia8_Sim08h_Digi13_Trig0x40760037_Reco14c_Stripping20r1NoPrescalingFlagged.root`

### Folder names
Proposal is `YYMMDD-<particle_name>-<reconstruction_mode>`. My comments
- I think we should combine productions by date so that we don't have too many top folders
- For me it is much easier to read the date if we write it as YY_MM_DD
- I'd add some description of the production

An example name with my suggestions may look as `20_05_08-unstripped/Dst-mc`

**Comment by yipengsun** [comment]
I think the `YY_MM_DD` date looks better, and we can stick with that. I guess putting the reconstructed particles first also makes sense, since this makes auto-completion easier. 

For the `--` separator, I think it's way too long. We should try to make the filename as short as possible. A filename that is too long is very hard to decipher. I think your example filename is way too long. This will be more problematic when using `ls` instead of `ls -l`.

So the revised filename would look like this:
```
Dst-20_02_20-mc-2012-md-dv36-py6-sim08a-Bd2DstTauNu.root
```

Also, I don't like the idea of combining productions. I don't think one folder per prodcution-date is excessive.

**Comment by yipengsun** [comment]
For the wrong-sign particles, since they are reconstructed with the Dst/D0, I think we don't need to mark them explicitly in the filename.

In the future, if/when we combine Dst/D0 reconstruction into a single script, the `particle_name` can just be `Dst_D0`.

**Comment by manuelfs** [comment]
Many groups at CMS used long names without problems, and they allowed us to identify the sample accurately. I worry about mistakes being made when transcribing the name to the shortened form and in the bookkeeping that says which sample that is. But if you commit to keep that well documented and up to date, you can use the shorter names. But do add the MC ID at least.

For the folders, we need to keep in one folder all the ntuples that are produced with the same version of the code and are meant to be used together. For instance, if the `D0` and `Dst` samples should be together because they are part of the same fit, and when you copy the ntuples to your laptop for that, we should be copying just one folder.

**Comment by yipengsun** [comment]
For the MC ID, do you mean something like `12573000`? That works if the ntuple contains only ONE mode, but in our case, some of Phoebe's ntuples have multiple modes combined. 

Also, the actual _decay mode_ <-> _MC ID_ correspondence are recorded in the `ganga_jobs.py` script. So I think my original naming (write down the decay mode `Bd2DstTauNu`, instead of the MC ID) still works.

I think if the naming convention is too hard to follow, then few is going to follow it. Also, I think for the `0.9` release, I'll include a Python script that checks the naming convention of selected files, so at least we know if the filenames are legal.

As for the one-folder-per-version strategy, I think we can enforce that starting from `0.8.6`. It certainly makes sense to only submit GRID jobs with a tagged version. So the folder names would look like this:
```
pre0.8.6-Dst_D0-data,mc,cutflow_data,cutflow_mc
0.8.6-Dst-cutflow_mc
```

No date will be included in the folder. The date will be included in the ntuple filename only.
```

**Comment by yipengsun** [comment]
BTW, I think we should include the reconstruction mode in the ntuple filename:
```
Dst--20_05_08--cutflow_mc--MC_2011_11874091_Beam3500GeV-2011-MagDown-Nu2-Pythia8_Sim08h_Digi13_Trig0x40760037_Reco14c_Stripping20r1NoPrescalingFlagged.root
```

**Comment by manuelfs** [comment]
I updated the notation
https://umd-lhcb.github.io/lhcb-ntuples-gen/ntupling/nomenclature/

Is there a way to preview the markdown for that specific wiki? I don't know how to fix the first example, but since we may change more things, I let it be. Let me know if you have comments

**Comment by yipengsun** [comment]
Suggestions for the naming of the folder structure:
```
0.8.6-cutflow/Dst-cutflow_mc
0.8.6-cutflow/Dst-cutflow_data
```

Instead of just `Dst-mc` or `Dst-data`. Because `cutflow_mc` is generated with a different configuration and should not be confused with regular MC. Also, I don't think when you just need regular MC ntuples, you want to copy the cutflow MC ntuples at the same time.

**Comment by manuelfs** [comment]
Yeah, it all depends on actual use. Usually, the type of MC is unique to the commit, so `0.8.6-cutflow` in the main folder would uniquely identify it.

If we have several possible tags with a given commit, then they would need to be separated in the inner folders, but we may be able to avoid this.

**Comment by yipengsun** [comment]
Also, can we use `YY_MM_DD`, instead of `YY-MM-DD`? Because the first date format can be used for log files, where the field separator is a single `-`, instead of a double `--`.

**Comment by yipengsun** [comment]
I plan to enforce naming convention in `0.8.6`, and finish the automatic check in `0.9`.

**Comment by yipengsun** [comment]
For the log files and cond files, I'll use `-` as separator. The `--` separator will only be used for all ntuples.

**Comment by yipengsun** [comment]
I made small updates to the naming convention. @manuelfs please take another look.
